SparsityPattern: Fix copy_from a DynamicSparsityPattern

tamiko opened the pull request:
Fix a logic issue in DynamicSparsityPattern::exists

If the row index i is not an element of the supplied (nonzero) rowset, then
the entry (i, j) does not exist. This is a perfectly valid situation that
might occur in case a DynamicSparsityPattern gets constrained to a (locally
relevant) index set in a distributed computation.

Without this fix, it is currently not possible to copy_from a
DynamicSparsityPattern into a static SparsityPattern. (In debug mode due to
an Assert triggering, in release mode due to an out of bounds read.)

tamiko commented:
/rebuild

bangerth commented:
I'm not in love with this solution since it returns wrong information -- the entry *does* exist, we just don't know about it on the current processor. My preference would be if we just didn't call the function in question when it can't compute its answer. What is the call chain that produces your error?

tjhei commented:
> Without this fix, it is currently not possible to copy_from a
> DynamicSparsityPattern into a static SparsityPattern.

Why do you want to copy from a DSP with a local range that is not the whole set into a SP? The problem is that ``copy_from()`` should have a special implementation for when the DSP doesn't store all rows (and then for example only queries those existing rows). That said, why do you do this in the first place and how is this useful? :-)

tamiko commented:
@tjhei  @bangerth 
I am developing an explicit compressible Euler solver that fundamentally operates over the stencil of a finite element space. For this, I have to store precomputed offline data per computing node locally for every stencil entry. I want to use a `dealii::SparseMatrix<double>` for this because I want to avoid implicit MPI communication, in essence this is somehwat a "matrix free" (or at least Trilinos/Petsc free) approach.

For this to work, two things have to be addressed:
 - It is currently not possible to `copy_from` a dynamic sparsity pattern into a (static) sparsity pattern used for the deal.II matrix.
 - The (static) sparsity pattern needs a slightly modified data structure to avoid creating diagonal entries unconditionally. (Was there ever a rationale behind this exceptthat this is somewhat a design by accident?)
 
@bangerth 
We already have `DynamicSparsityPattern::row_length(i)` that returns 0 if the IndexSet is unaware of the index in question instead of throwing an error (line 1069 of dynamic_sparsity_pattern.h). Why should `DynamicSparsityPattern::exists` behave differently?

bangerth commented:
On 4/15/19 3:24 PM, Matthias Maier wrote:
> <https://github.com/bangerth>
> We already have |DynamicSparsityPattern::row_length(i)| that returns 0 
> if the IndexSet is unaware of the index in question instead of throwing 
> an error (line 1069 of dynamic_sparsity_pattern.h). Why should 
> |DynamicSparsityPattern::exists| behave differently?

I would claim that the former is the mistake, not the latter. (And from 
his answer, I suspect that @tjhei also agrees with this.)

We're actually pretty consistent in using assertions to make sure 
callers don't enquire information we don't know on a processor, rather 
than returning answers that seem correct but may not (and for which the 
caller has no way of figuring out whether it's correct).

tjhei commented:
> * It is currently not possible to `copy_from` a dynamic sparsity pattern into a (static) sparsity pattern used for the deal.II matrix.

Okay, would it make sense to amend the ``copy_from`` to only iterate over the existing rows then?

bangerth commented:
On 4/15/19 3:27 PM, Timo Heister wrote:
>       *
> 
> Okay, would it make sense to amend the |copy_from| to only iterate over 
> the existing rows then?
> 

Yes.

tamiko commented:
@bangerth I understand your argument, but I disagree. The `IndexSet` for a `DynamicSparsityPattern` restricts the range of row indices the sparsity pattern operates on. This is an operation that is agnostic how it is actually used. And actually everything in `DynamicSparsityPattern` silently drops entries without throwing an assertion except `exists`.

On the other hand, I am a bit confused by the following: When you create a `DynamicSparsityPattern` say over the locally relevant indices for a processor, then you will end up with incomplete lines in your stencil (beause you miss part of the neighborhood of a locally relevant but not locally owned degree of freedom). So the information you get by reading out a line is already a local view and not a global one. So why not taking the obvious stance and saying that if a row is ignored (due to the index set) then its local size is 0 - we report possible "incorrect" (or better "inclomplete") answers for all other rows anyway.

tamiko commented:
/rebuild

tamiko commented:
/rebuild

tjhei commented:
I think you will need to fix the output of ``sparsity/dynamic_sparsity_pattern_21.debug``, just rebuilding won't fix it. :-)

tamiko commented:
*ugh* Somehow github isn't picking up my force push to the branch ?!

tamiko commented:
So, now we're talking :-) (I pushed an empty commit which triggered github to pick up the changes.)

tamiko commented:
Let's wait for reaching some consensus on this pull request :-)

tamiko commented:
@dealii/dealii How shall we proceed?

bangerth commented:
@tamiko: The difference in perspective is whether you consider the object on a given processor as an entity in itself, or as a part of a larger, distributed data structure. 

The approach we have widely taken in deal.II is to think of these data structures as part of a larger something. In all of these cases, when a local object cannot report on something that may exist, but is not available on the local processor, then you get an error. Examples are: Asking for DoF indices on artificial cells; asking for the active fe index on artificial cells; asking for the entries of a matrix row or a vector entry for entries not locally store.

The philosophy behind throwing an exception is that calling a function asking for information that is not available likely suggests a programming error. It probably indicates that the programmer didn't think things through completely, or that the algorithm works for sequential programs but has never been converted for the parallel case. In all of these situations, the right approach is to really think about the correctness of the algorithm that made the wayward function call. What you are suggesting is to let the function return a value that may look correct, but likely is not. 

For why this is dangerous: Think about making decisions based on the number of entries in a matrix. Every processor in a program originally written for the sequential case now loops over all rows of the matrix, adds up the number of entries per row, and then makes decisions. If you do this in parallel and the `row_length()` function just returns zero for rows not locally stored, then your algorithm as written produces a number that, for all practical purposes, looks like the right one ("We did loop over all rows of the matrix!") but it is not; worse, every processor now stores a different number for something that really looks like it should be the same -- think about the difficulty of debugging something like this if you didn't want to single-step into the `row_length()` function!

So for all of these reasons, the defensive programming approach is to throw an exception if a function can't figure out the answer to a question. At least as long as the *calling* site has a way to figure out whether the called function can answer the question. In the current case, the calling site can, because it knows (or can get access to) the index set. I believe that this is the right philosophy to build good software on. (And if there are functions that don't currently follow this scheme, then these are problematic and should be changed, rather than considered as good examples.)

tamiko commented on source/lac/dynamic_sparsity_pattern.cc:
```diff
@@ -363,6 +363,11 @@ DynamicSparsityPattern::exists(const size_type i, const size_type j) const
+  // Avoid a segmentation fault in below code if the row index happens to
+  // not be present in the IndexSet rowset:
+  if (rowset.size() != 0 && !rowset.is_element(i))
+    return false;
```
@bangerth I have left the Assert that you strongly insist on in place. I suggest to still merge the 4 lines in order to avoid a segmentation fault.

bangerth replied:
In other words, if you make it past the assertion, then `rowset.is_element(i)==true` and the condition here can never be satisfied. Do I understand right then that you're only trying to avoid the segfault that would happen in release mode but that could never happen in debug mode?

tamiko replied:
Yes.

bangerth replied:
Then maybe write it like this:
```
  // The assertion above guarantees that we are dealing with an 'i' for which we store rows.
  // But if one runs the code below in release mode, one would get a segfault, so be friendly
  // and avoid this:
  if (! (...condition in assertion...))
    return 0;
```
Though I'm not sure we should really be friendly to people who run in release mode and complain about segfaults -- they get what they deserve.

tamiko replied:
@bangerth Since when are we so mean to users?

tamiko commented on source/lac/sparsity_pattern.cc:
```diff
@@ -494,7 +494,7 @@ SparsityPattern::copy_from(const DynamicSparsityPattern &dsp)
     {
       row_lengths[i] = dsp.row_length(i);
-      if (do_diag_optimize && !dsp.exists(i, i))
+      if (do_diag_optimize && (dsp.row_length(i) == 0))
```
This as well fixes the read. An alternative approach would be to iterate over the IndexSet instead.

bangerth replied:
Do you have access to the index set of the DSP? If so, that might be the better way to test whether a line is locally stored.

bangerth commented on source/lac/dynamic_sparsity_pattern.cc:
```diff
@@ -358,11 +358,20 @@ DynamicSparsityPattern::exists(const size_type i, const size_type j) const
+    rowset.size() == 0 || rowset.is_element(i),
+    ExcMessage(
+      "The row IndexSet does not contain the index i. This sparsity pattern
+      object cannot know whether the entry (i, j) exists or not."));
```
nice!

tamiko commented:
@bangerth @masterleinad I think I have addressed all comments on this one. Please review (I would like to see this getting merged rather sooner than later.)

bangerth commented:
Looks good, thanks!
